### PR TITLE
MWPW-136488 Bulk Update Reporting

### PR DESCRIPTION
**Pull request by Brandon32**
* Updated migration functions to return status and message, extra data can be added dynamically. 
* Refactored blog migration to create reports.
* Updated fetch utils and docx utils.
* Created migration utils to account for both types of migrations, block and path.
* Removed cache and force parameters and added source parameter.

Resolves: [MWPW-136488](https://jira.corp.adobe.com/browse/MWPW-136488)

Report: https://main--bacom-blog--adobecom.hlx.page/drafts/staged-content/migration-report.json

### CODE REVIEW

**Review by meganthecoder — APPROVED** [review]
LGTM

File: blog-migration.js [hunk 1]
```diff
@@ -13,206 +13,302 @@
...
+    if (blockMigrations.length > 0) {
+        const blocks = blockMigrations.map(([block]) => block).join(', ');
+        const blockReport = await migrateBlocks(mdast, blockMigrations, entry);
+
+        // Save docx
+        if (blockReport.status.status === STATUS_FAILED) {
+            console.warn(`${pageIndex} migration ${index} ${STATUS_FAILED} '${entry}' - blocks: '${blocks}'`);
+            console.warn(`${pageIndex} ${STATUS_FAILED}: '${blockReport.status.message}'`);
+            blockReport.status.save = STATUS_FAILED;
+            blockReport.status.saveMessage = 'Migration failed, skipping save'
```
Comment (JasonHowellSlavin) [inline blog-migration.js:193]:
```suggestion
            blockReport.status.saveMessage = 'Migration failed, skipping save';
```

File: blog-migration.js [hunk 2]
```diff
@@ -13,206 +13,302 @@
...
+        process.exit(1);
+    }
 
-            report.succeed.push({ entry, migratedBlocks: migrationBlocks });
-        } catch (e) {
-            console.error(`Error migrating ${entry}`, e.message);
-            report.failed.push({ entry, message: e.message });
+    for (let i = 0; i < entries.length; i++) {
+        const entry = entries[i].trim();
+        const url = `${FROM_SITE}${entry}`
```
Comment (JasonHowellSlavin) [inline blog-migration.js:284]:
```suggestion
        const url = `${FROM_SITE}${entry}`;
```

File: utils/migration-utils.js [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+ * @param {Array<string, function>} migrationMap - paths to migrate
+ * @param {string} entry - entry path
+ * @returns {Promise<Array>} - reports for each page
+ */
+export const migratePaths = async (mdast, migrationMap, entry, sourceDocxFile, outputDocxFile) => {
+    const reports = [];
+
+    for (const [path, migrate] of migrationMap) {
+        // Migrate to a new page
+        const key = `path ${path.replaceAll('/', ' ').trim()}`
```
Comment (JasonHowellSlavin) [inline utils/migration-utils.js:193]:
```suggestion
        const key = `path ${path.replaceAll('/', ' ').trim()}`;
```

File: bacom-blog/tag-header/tag-header.js [hunk 4]
```diff
@@ -14,12 +20,12 @@ export async function convertTagHeader(mdast) {
 
     const image = select('paragraph', mdast);
     if (!image) {
-        return 'No image present in tag header'
+        return { status: STATUS_WARNING, message: 'No image present in tag header'}
```
Comment (JasonHowellSlavin) [inline bacom-blog/tag-header/tag-header.js:23]:
```suggestion
        return { status: STATUS_WARNING, message: 'No image present in tag header'};
```

File: bacom-blog/tag-header/tag-header.js [hunk 5]
```diff
@@ -14,12 +20,12 @@ export async function convertTagHeader(mdast) {
...
     }
 
     const heading = select('heading', mdast);
     if (!heading) {
-        return 'No heading present in tag header'
+        return { status: STATUS_WARNING, message: 'No heading present in tag header'}
```
Comment (JasonHowellSlavin) [inline bacom-blog/tag-header/tag-header.js:28]:
```suggestion
        return { status: STATUS_WARNING, message: 'No heading present in tag header'};
```

File: blog-migration.js [hunk 6]
```diff
@@ -13,206 +13,302 @@
...
-    const outputDir = `${output}/${PROJECT}`;
-    const report = { succeed: [], skipped: [], failed: [], warned: [] };
-    const entries = await readIndex(index);
-    Object.keys(MIGRATION).map((key) => report[key] = []);
+function formatReportData(reports) {
+    const statusReports = {};
+    const migrationReports = {};
+    const PAGE_TOTAL = 'page total';
+    const MIGRATION_TOTAL = 'migration total';
+    const totals = { [PAGE_TOTAL]: [], [MIGRATION_TOTAL]: [] }
```
Comment (JasonHowellSlavin) [inline blog-migration.js:97]:
```suggestion
    const totals = { [PAGE_TOTAL]: [], [MIGRATION_TOTAL]: [] };
```

**Review by JasonHowellSlavin — APPROVED** [review]
Looks great! Small suggestions, nothing consequential.

File: bacom-blog/aside/aside.js [hunk 7]
```diff
@@ -26,5 +31,17 @@ export async function bannerToAside(mdast) {
...
+    const fragmentDocxFile = outputDocxFile.replace(BANNERS_PATH, FRAGMENTS_PATH);
+    report.output = fragmentDocxFile;
+
+    if (image) {
+        report.status = STATUS_SUCCESS;
+        report.message = 'Banner converted to aside with image';
+    } else {
+        report.status = STATUS_SUCCESS;
+        report.message = 'Banner converted to aside without image';
+    }
```
[the remark this answers is not in the document]
Comment (Brandon32) [reply bacom-blog/aside/aside.js]: Updated

File: blog-migration.js (hunk 1 above)
Comment (Brandon32) [reply blog-migration.js]: Fixed

File: blog-migration.js (hunk 2 above)
Comment (Brandon32) [reply blog-migration.js]: Fixed

File: utils/migration-utils.js (hunk 3 above)
Comment (Brandon32) [reply utils/migration-utils.js]: Fixed

File: bacom-blog/tag-header/tag-header.js (hunk 4 above)
Comment (Brandon32) [reply bacom-blog/tag-header/tag-header.js]: Fixed

File: bacom-blog/tag-header/tag-header.js (hunk 5 above)
Comment (Brandon32) [reply bacom-blog/tag-header/tag-header.js]: Fixed

File: blog-migration.js (hunk 6 above)
Comment (Brandon32) [reply blog-migration.js]: Fixed

**Pull request merged**